### PR TITLE
wgengine: start logging DISCO frames to pcap stream

### PR DESCRIPTION
**Pull request by twitchyliquid64**

### CODE REVIEW

**Comment by twitchyliquid64** [comment]
image

File: wgengine/capture/capture.go [hunk 1]
```diff
@@ -65,6 +65,9 @@ const (
 	// SynthesizedToPeer indicates the packet was generated from within tailscaled,
 	// and is being routed to a remote Wireguard peer.
 	SynthesizedToPeer Path = 3
+
+	// PathDisco indicates the packet is metadata about a valid disco frame.
```
Comment (bradfitz) [inline wgengine/capture/capture.go:69]: drop "valid"? invites more questions. (like: is there a path for invalid disco frames?)

And is it just "metadata"? It also contains the payload. Maybe you want something like "a decrypted disco message"?

File: wgengine/magicsock/magicsock.go [hunk 2]
```diff
@@ -80,6 +82,12 @@ const (
 	socketBufferSize = 7 << 20
 )
 
+// CaptureFunc describes a callback to record packets when
+// debugging packet-capture. Such callbacks must not take
+// ownership of the provided data slice: it may only copy
+// out of it within the lifetime of the function.
+type CaptureFunc func(capture.Path, time.Time, []byte)
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock.go:89]: why's this here in magicsock? if you're already importing the capture package?

File: wgengine/magicsock/magicsock.go [hunk 3]
```diff
@@ -348,6 +356,9 @@ type Conn struct {
 	// stats maintains per-connection counters.
 	stats atomic.Pointer[connstats.Statistics]
 
+	// pcap logging callback - non-nil when capturing.
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock.go:359]: // captureHook, if non-nil, is the pcap logging callback when capturing.

File: wgengine/magicsock/magicsock.go [hunk 4]
```diff
@@ -664,6 +675,10 @@ func NewConn(opts Options) (*Conn, error) {
 	return c, nil
 }
 
+func (c *Conn) InstallCaptureHook(cb CaptureFunc) {
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock.go:678]: docs

File: wgengine/magicsock/magicsock.go [hunk 5]
```diff
@@ -2099,6 +2114,29 @@ func (c *Conn) handleDiscoMessage(msg []byte, src netip.AddrPort, derpNodeSrc ke
 		return
 	}
 
+	// Emit information about the disco frame into the pcap stream
+	// if a capture hook is installed.
+	if cb := c.captureHook.Load(); cb != nil {
+		var (
+			l    bytes.Buffer
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock.go:2121]: never use `l` as a variable name. Looks terrible in too many fonts. :)

File: wgengine/magicsock/magicsock.go [hunk 6]
```diff
@@ -2099,6 +2114,29 @@ func (c *Conn) handleDiscoMessage(msg []byte, src netip.AddrPort, derpNodeSrc ke
 		return
 	}
 
+	// Emit information about the disco frame into the pcap stream
+	// if a capture hook is installed.
+	if cb := c.captureHook.Load(); cb != nil {
+		var (
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock.go:2120]: pull this all out into a func? handleDiscoMessage is long enough already.

File: wgengine/magicsock/magicsock.go [hunk 7]
```diff
@@ -2099,6 +2114,29 @@ func (c *Conn) handleDiscoMessage(msg []byte, src netip.AddrPort, derpNodeSrc ke
...
+			l    bytes.Buffer
+			flag uint8
+		)
+		if src.Addr() == derpMagicIPAddr {
+			flag |= 0x01
+		}
+		l.Write([]byte{flag}) // 1b: flag
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock.go:2127]: WriteByte?

File: wgengine/magicsock/magicsock.go (hunk 5 above)
Comment (bradfitz) [inline wgengine/magicsock/magicsock.go:2121]: call Grow on it first, before all the little writes?

but this is garbage per disco message. probably okay. Maybe just add a comment like "// Warning: alloc garbage. Acceptable when capturing."

**Review by bradfitz — COMMENTED** [review]
Exciting!

File: wgengine/magicsock/magicsock.go (hunk 2 above)
Comment (twitchyliquid64) [reply wgengine/magicsock/magicsock.go:89]: Just keeping the type where its used. We already have equivalents in tstun and wgengine, if we want to centralize it maybe we should rip the same out of there too?

Comment (bradfitz) [reply wgengine/magicsock/magicsock.go:89]: yeah, one place would be ideal

**Review by bradfitz — APPROVED** [review]

**Pull request merged**